### PR TITLE
Add TupleNVarULE

### PR DESCRIPTION
**Pull request by Manishearth**
The VarULE counterpart of TupleNULE

Part of https://github.com/unicode-org/icu4x/issues/5523. Planned to be used in https://github.com/unicode-org/icu4x/issues/4437

I'm not super happy with the naming with this vs VarTupleULE, but I've tried to make it clearer with the module names and it's fine for now. We can rename as desired since zerovec isn't on the ICU4X stability track.

I do plan to add serde/etc impls but that's going to be a separate PR.

### CODE REVIEW

**Comment by Manishearth** [comment]
Hm, my serde impls here aren't optimal for postcard, I think

File: utils/zerovec/src/ule/tuplevar.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        impl<'a, $($T,)+ $($T_alt),+> ZeroFrom <'a, $name<$($T,)+>> for ($($T_alt),+)
+        where
+                    $($T: VarULE + ?Sized,)+
+                    $($T_alt: ZeroFrom<'a, $T>,)+ {
+            fn zero_from(other: &'a $name<$($T,)+>) -> Self {
+                (
+                    $($T_alt::zero_from(other.$t()),)+
+                )
+            }
+        }
```
Comment (sffc) [inline utils/zerovec/src/ule/tuplevar.rs:169]: Nit: I think better would be a non-trait associated function `to_tuple(&self)` that returns `(&T1, &T2, ...)`

File: utils/zerovec/src/ule/tuplevar.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+        #[cfg(feature = "serde")]
+        impl<$($T: serde::Serialize),+> serde::Serialize for $name<$($T),+>
+        where
+            $($T: VarULE + ?Sized,)+
+            // This impl should be present on almost all VarULE types. if it isn't, that is a bug
+            $(for<'a> &'a $T: ZeroFrom<'a, $T>),+
```
Comment (sffc) [inline utils/zerovec/src/ule/tuplevar.rs:176]: Question/Issue: `ZeroFrom` is intended to be implemented on sized types. You shouldn't need it for a reference; just take the `&T` reference directly.

File: utils/zerovec/src/ule/tuplevar.rs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: serde::Serializer {
+                let this = <($(&$T),+)>::zero_from(self);
+                <($(&$T),+) as serde::Serialize>::serialize(&this, serializer)
+            }
+        }
+
+        #[cfg(feature = "serde")]
+        impl<'de, $($T: VarULE + ?Sized),+> serde::Deserialize<'de> for Box<$name<$($T),+>>
```
Comment (sffc) [inline utils/zerovec/src/ule/tuplevar.rs:185]: Observation: you are missing the third serde impl, but you noted that you'll be following up with another PR to add the rest of the serde impls.

Maybe remove the serde impls from this PR?

File: utils/zerovec/src/ule/tuplevar.rs (hunk 1 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/tuplevar.rs:169]: The trait is needed for other traits to compose

File: utils/zerovec/src/ule/tuplevar.rs (hunk 3 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/tuplevar.rs:185]: No, I just changed the plan to include serde impls here when I realized it's all a bit linked.

And yeah I plan to add the third impl

Comment (Manishearth) [reply utils/zerovec/src/ule/tuplevar.rs:185]: I can split the impls out into a separate PR.

File: utils/zerovec/src/ule/tuplevar.rs (hunk 1 above)
Comment (sffc) [reply utils/zerovec/src/ule/tuplevar.rs:169]: ok, the impl seems harmless.

We might still want `to_tuple`.

**Review by sffc — APPROVED** [review]
